Implement first set of federated service e2e tests.

madhusudancs opened the pull request:
These tests are untested and there is no guarantee that they work. The ongoing auth problems is blocking these e2es from being tested and upon @quinton-hoole's request I am submitting them now.

Only the last commit here needs review.

Depends on #26953

cc @nikhiljindal @colhom @mfanjie @kubernetes/sig-cluster-federation

colhom commented on federation/cluster/common.sh:
```diff
@@ -24,7 +24,14 @@
+detect-project
+
 FEDERATION_IMAGE_REPO_BASE=${FEDERATION_IMAGE_REPO_BASE:-'gcr.io/google_containers'}
+FEDERATION_PUSH_REPO_BASE=${FEDERATION_PUSH_REPO_BASE:-"gcr.io/${PROJECT}"}
```
wonderful! thanks for doing that.

goltermann commented:
@madhusudancs When do you think you'll have this one ready for review?  We cut beta on June 10.

madhusudancs committed:
Implement first set of federated service e2e tests.

Note that these tests are untested and there is no guarantee that they work.
The ongoing auth problems is blocking these e2es from being tested and upon
@quinton-hoole's request I am submitting them now.

madhusudancs commented:
@goltermann it is ready now.

colhom commented:
@madhusudancs e2e is failing due to #23545

madhusudancs commented:
@colhom Thanks for pointing it out.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+		}
+
+		By("Obtaining a list of all the clusters")
+		clusterList, err := f.FederationClientset.Federation().Clusters().List(api.ListOptions{})
```
I think you need a check here that the number of clusters matches the number of contexts above, with a few retries on the list operation, given that the cluster create call is theoretically asynchronous.

quinton-hoole replied:
Also, given that you register/create cluster objects in BeforeEach() (i.e. before every test, i.e. before every It() clause), you also need to unregister/delete the cluster objects after every test (i.e. in AfterEach()).  Either that, or you need to register the clusters once before all the tests, and delete them once after all the tests (i.e in BeforeSuite() and AfterSuite()) See https://onsi.github.io/ginkgo/#global-setup-and-teardown-beforesuite-and-aftersuite

madhusudancs replied:
Done. Didn't put it in Before/AfterSuite because that would run the setup for all the tests.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+	})
+
+	It("should be able to discover a federated service", func() {
+		framework.SkipUnlessFederated(f.Client)
```
You should also skip all of the BeforeEach() stuff above (cluster registration etc) unless we're federated.

madhusudancs replied:
Done.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+	})
+})
+
+func waitForFederatedServiceShard(cs *release_1_3.Clientset, namespace string, service *api.Service, numSvcs int) {
```
minor nit: Probably worth adding a brief comment describing what this function does.

madhusudancs replied:
Done.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+func waitForFederatedServiceShard(cs *release_1_3.Clientset, namespace string, service *api.Service, numSvcs int) {
+	By("Fetching a federated service shard")
+	var clSvcList *v1.ServiceList
+	for t := time.Now(); time.Since(t) < FederatedServiceTimeout; time.Sleep(framework.Poll) {
```
nit: There are some library functions (wait*) in our test libraries that will do this for you.  Better to use the standard libraries than roll your own retry/wait logic here.

quinton-hoole replied:
k8s.io/kubernetes/pkg/util/wait is what you're looking for.  This can wait for a followup PR if you prefer.

madhusudancs replied:
Done.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+		if n == numSvcs {
+			break
+		}
+		framework.Logf("%d services found, trying again in %s", n, framework.Poll)
```
nit: It would be good to add the number of services that you're waiting for in this log message.

madhusudancs replied:
Done.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+	for t := time.Now(); time.Since(t) < FederatedServiceTimeout; time.Sleep(framework.Poll) {
+		var err error
+		clSvcList, err = cs.Core().Services(namespace).List(api.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
```
Does this succeed even when the list of services is zero length the first time?

quinton-hoole replied:
Never mind.  I think it's fine, as long as listing services when none exist does not return an error (which it shouldn't).

madhusudancs replied:
Ack.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+	Expect(len(clSvcList.Items)).To(Equal(numSvcs))
+	if numSvcs > 0 {
+		// Renaming for clarity/readability
+		clSvc := clSvcList.Items[0]
```
This only seems to work for numSvcs == 1 ?  The semantics of this function are rather confusing.  As far as I can tell it will wait until the number of services in the specifed namespace reaches numSvcs, and (presumably this only works if 'service' is not nil) if numSvcs >0, that the first listed service matches the one passed in?  You should definitely document the semantics in the header, and also check for service != nil before dereferencing it.

madhusudancs replied:
> This only seems to work for numSvcs == 1 ?

That's right and that's intentional. It only makes sense to execute the block inside the `if` only if we expect to have at least one shard.
 
> The semantics of this function are rather confusing. As far as I can tell it will wait until the number of services in the specifed namespace reaches numSvcs, and (presumably this only works if 'service' is not nil) if numSvcs >0, that the first listed service matches the one passed in? You should definitely document the semantics in the header, and also check for service != nil before dereferencing it.

Thanks for catching that. Added a check for service != nil. Also added the doc string for the function.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+	}
+}
+
+func createService(fcs *federation_internalclientset.Clientset, clusterClientSets []*release_1_3.Clientset, ns string) {
```
nit: prefer 'namespace' rather than 'ns'

madhusudancs replied:
Done.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+	}
+}
+
+func discoverService(f *framework.Framework) {
```
You should possibly take the DNS name to lookup as a parameter here, and call this more than once looking up more than one of the service's DNS names? e.g. myservice, myservice.mynamespace, myservice.mynamespace.myfederation etc.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+				{
+					Name:    "federated-service-discovery-container",
+					Image:   "gcr.io/google_containers/busybox:1.24",
+					Command: []string{"sh", "-c", "nslookup", FederatedServiceName},
```
You'll only end up looking up myservice.mynamespace.cluster.local here, and not the federated service?

madhusudancs replied:
I am not sure if I understand the question correctly. Can you please rephrase/elaborate?

madhusudancs replied:
I see what you mean now. Thanks for catching that. Fixed.

quinton-hoole commented:
Thanks @madhusudancs first pass of review done.  Tomorrow I will try to patch this in and run it against our standard e2e test setup.

dchen1107 commented:
This one needs a rebase ...

nikhiljindal commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,209 @@
+/*
```
name this federation-service?

madhusudancs replied:
This shouldn't be renamed to "federation-service" like the way we call other items that belong to the federation. This service is "federated" across a federation, a verb, in addition to being part of the federation.

madhusudancs committed:
Address review comments.

madhusudancs commented:
@quinton-hoole Addressed the comments, PTAL.

nikhiljindal commented:
Thanks @madhusudancs 
Can you please rebase now that https://github.com/kubernetes/kubernetes/pull/26636 is merged.

cc @mml who was trying to patch and run this.

@mml or @madhusudancs Have you guys been able to verify that these pass?

madhusudancs commented:
> Can you please rebase now that #26636 is merged.

@nikhiljindal Which PR did you mean? This seems like a self-recursion :)

nikhiljindal commented:
Sorry I meant https://github.com/kubernetes/kubernetes/pull/26953.
The first commit in this PR is already submitted right?

madhusudancs commented:
Ah Ok. Thanks. Yes, I will rebase on top of that PR. It shouldn't affect trying this PR though, since that commit is included here.

I haven't been able to verify that the tests pass. I am trying to bring up a new federation based on the recent merges.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,283 @@
+// service reaches the expected value, i.e. numSvcs in the given individual Kubernetes
+// cluster. If the shard count, i.e. numSvcs is expected to be at least one, then
+// it also checks if the first shard's name and spec matches that of the given service.
+func waitForFederatedServiceShard(cs *release_1_3.Clientset, namespace string, service *api.Service, numSvcs int) {
```
nit: I don't understand how one could ever have more than 1 instance of the service in a given cluster at the same time.  There would be a name clash, right?  In which case, is the numSvcs parameter essentially a boolean?  Would it ever make sense to call this function with that parameter not equal to either 0 or 1?

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,283 @@
+	var clSvcList *v1.ServiceList
+	if err := wait.PollImmediate(framework.Poll, FederatedServiceTimeout, func() (bool, error) {
+		var err error
+		clSvcList, err = cs.Core().Services(namespace).List(api.ListOptions{})
```
Don't you only want to list the service with the given name (service.Name()) here?  I realize that this test should run in it's own namespace, and that there should be no other services in that namespace, but the logic in this function still seems overly complex and confusing to me. 

Should you not simply be listing the services with the given name, and confirming that the length of the returned list is either 0 or 1 (depending on the value of numServices, which I suspect should be a bool, not an int).

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,283 @@
+		framework.Logf("%d services found, waiting for %d, trying again in %s", n, numSvcs, framework.Poll)
+		return false, nil
+	}); err != nil {
+		framework.Failf("Failed to list registered clusters: %+v", err)
```
You're listing services here, not clusters.  Cut-n-paste error?

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,283 @@
+	}
+	_, err := fcs.Core().Services(namespace).Create(service)
+	Expect(err).NotTo(HaveOccurred())
+	for _, cs := range clusterClientSets {
```
You need to add a deferred function here to delete the created service.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,283 @@
+	FederatedServicePod  = "federated-service-test-pod"
+
+	// TODO: Only suppoprts IPv4 addresses. Also add a regexp for IPv6 addresses.
+	FederatedIPAddrRegexp  = `(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])`
```
nit: Woudn't it be better to use a simpler regex like [0-9]{1,3}\.){4}

quinton-hoole replied:
nit: This is not specific to anything federated.  Would it not be better to call it something like IPv4Regexp and put it in the utils library for other people to also use?

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,283 @@
+
+	// TODO: Only suppoprts IPv4 addresses. Also add a regexp for IPv6 addresses.
+	FederatedIPAddrRegexp  = `(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])`
+	FederatedDNS1123Regexp = `([a-z0-9]([-a-z0-9]*[a-z0-9])?\.)*([a-z0-9]([-a-z0-9]*[a-z0-9])?)`
```
nit: Similar comment to above.

quinton-hoole commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,283 @@
+	}
+	if exists {
+		f.TestContainerOutputRegexp("federated service discovery", pod, 0, []string{
+			`Name:\s+` + FederatedDNS1123Regexp + `\nAddress \d+:\s+` + FederatedIPAddrRegexp + `\s+` + FederatedDNS1123Regexp,
```
nit: Surely you want the actual name of the service in FederatedDNS1123Regexp?  Otherwise you'll match any DNS name that comes back.  I guess that doesn't really matter in practise.  To make this test more robust I think it would be better to only distinguish between DNS lookup success and failure.  The exact output and format of the nslookup tool is mostly irrelevant, and might change over time and/or platform.  I checked the nslookup tool and unfortunately it returns zero on both success and failure.  

But the 'host' tool is a little better for this.  It is pretty ubiquitous, and prints the name and address on success, but prints nothing on failure. That might be a simpler, more robust test.

```
$ host www.google.com
www.google.com has address 172.217.2.228
$ host rubbish
$
```

quinton-hoole commented:
Thanks @madhusudancs .  All my remaining comments can be addressed in a followup PR.  Please do.  LGTM for this PR.

nikhiljindal commented on test/e2e/federated-service.go:
```diff
@@ -0,0 +1,283 @@
+		framework.SkipUnlessFederated(f.Client)
+
+		// TODO: Federation API server should be able to answer this.
+		if federationName = os.Getenv("FEDERATION_NAME"); federationName == "" {
```
No one is setting this env var anywhere?
So the tests right now will definitely fail?

madhusudancs replied:
Yeah, I wasn't expecting this PR to merge before waking up.

nikhiljindal replied:
I have https://github.com/kubernetes/kubernetes/pull/27333 to fix this

madhusudancs replied:
Thanks!
